### PR TITLE
feat: introduce external plugin support in proxy

### PR DESCRIPTION
**Pull request by aylei**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [ ] Improvement
- [ ] Documentation
- [x] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

Part of https://github.com/matrixorigin/MO-Cloud/issues/369

## What this PR does / why we need it:

As https://github.com/matrixorigin/MO-Cloud/issues/369 suggested, this PR introduce external plugin support in proxy, which allows proxy users to extend the filtering and routing algorithm in a non-invasive way.

## Test

1. Unit test for regression;
2. New test cases are added to cover the behavior of `pluginRouter`;
3. manual test with an external plugin:

```bash
# enable plugin in proxy config
> echo '[proxy.plugin]\nbackend="127.0.0.1:32019"' >> ./etc/launch-with-proxy/proxy.toml
# start mo cluster
> ./mo-service -launch ./etc/launch-with-proxy/launch.toml -with-proxy

# tests
> mysql -h 127.0.0.1 -P6009 -udump -p
Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.
mysql>

# plugin block access of a tenant
> mysql -h 127.0.0.1 -P6009 -ublocked:root -p111
ERROR 1045 (28000): Auth failed: info: using a blocked Tenant

# plugin block access of an IP
> mysql -h 127.0.0.1 -P6009 -ublocked:root -p111
ERROR 1045 (28000): Auth failed: info: 127.0.0.1 is not allowed to access
```

### CODE REVIEW

File: pkg/pb/plugin/plugin.go
Comment (volgariver6) [inline pkg/pb/plugin/plugin.go:1]: add copy right and apache license header please.

File: pkg/proxy/config.go [hunk 1]
```diff
@@ -15,6 +15,9 @@
 package proxy
 
 import (
+	"context"
+	"github.com/matrixorigin/matrixone/pkg/common/moerr"
+	"github.com/matrixorigin/matrixone/pkg/util/errutil"
```
Comment (volgariver6) [inline pkg/proxy/config.go:20]: please seperate internal packages and external packages with a blank line.

File: pkg/proxy/plugin.go [hunk 2]
```diff
@@ -0,0 +1,126 @@
+package proxy
+
+import (
+	"context"
+	"github.com/matrixorigin/matrixone/pkg/common/moerr"
+	"github.com/matrixorigin/matrixone/pkg/common/morpc"
+	"github.com/matrixorigin/matrixone/pkg/logutil"
+	"github.com/matrixorigin/matrixone/pkg/pb/plugin"
```
Comment (volgariver6) [inline pkg/proxy/plugin.go:8]: please seperate internal packages and external packages with a blank line, and add copyright header.

File: pkg/proxy/plugin_test.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+import (
+	"context"
+	"errors"
+	"github.com/lni/goutils/leaktest"
+	"github.com/matrixorigin/matrixone/pkg/common/runtime"
+	"github.com/matrixorigin/matrixone/pkg/frontend"
+	"github.com/matrixorigin/matrixone/pkg/pb/metadata"
+	"github.com/matrixorigin/matrixone/pkg/pb/plugin"
+	"github.com/stretchr/testify/require"
+	"testing"
```
Comment (volgariver6) [inline pkg/proxy/plugin_test.go:12]: please seperate internal packages and external packages with a blank line, and add copyright header.

File: pkg/pb/plugin/plugin.go
Comment (aylei) [reply pkg/pb/plugin/plugin.go:1]: good catch, fixed

**Review by volgariver6 — APPROVED** [review]
LGTM

**Review by zhangxu19830126 — APPROVED** [review]

**Review by sukki37 — APPROVED** [review]

**Review by nnsgmsone — APPROVED** [review]

**Pull request merged**